### PR TITLE
Make SpecifyShape work for scalar shapes too

### PR DESCRIPTION
**Pull request by michaelosthege**
Also match exception messages in the tests.

### CODE REVIEW

File: aesara/tensor/shape.py [hunk 1]
```diff
@@ -371,22 +371,34 @@ class SpecifyShape(COp):
...
-        if shape.ndim > 1:
-            raise AssertionError()
-        if shape.dtype not in aesara.tensor.type.integer_dtypes:
-            raise AssertionError()
-        if isinstance(shape, TensorConstant) and shape.data.size != x.ndim:
-            raise AssertionError()
-        return Apply(self, [x, shape], [x.type()])
+        if shape == () or shape == []:
+            shape = np.array(shape, dtype=int)
+        tshape = aet.as_tensor_variable(shape)
```
Comment (brandonwillard) [inline aesara/tensor/shape.py:376]: Why can't we simply use `shape = aet.as_tensor_variable(shape, ndim=1)` here?

File: aesara/tensor/shape.py [hunk 2]
```diff
@@ -371,22 +371,34 @@ class SpecifyShape(COp):
...
+        if tshape.ndim != 1:
+            raise AssertionError(
+                f"The `shape` must be a vector. Got `shape={shape}` instead."
+            )
+        if tshape.dtype not in aesara.tensor.type.integer_dtypes:
+            raise AssertionError(
+                f"The `shape` must be an integer type. Got {tshape.dtype} instead."
+            )
```
Comment (brandonwillard) [inline aesara/tensor/shape.py:384]: We can just `cast` the `shape` to `int64`.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by michaelosthege** [comment]
The failing test is a regex pattern that doesn't match because the message differs between the Python/C perform.
Looking at the 🕝 I will fix this tomorrow, unless someone else does.

File: aesara/tensor/shape.py (hunk 1 above)
Comment (michaelosthege) [reply aesara/tensor/shape.py:376]: You're right, that's probably much better.
The option just wasn't on my radar.

File: aesara/tensor/shape.py (hunk 2 above)
Comment (michaelosthege) [reply aesara/tensor/shape.py:384]: If an element in a shape vector ever becomes a float, that should be an indication that something is (very) wrong upstream. So silently casting it (if we don't need to) didn't feel very nice.

Comment (brandonwillard) [reply aesara/tensor/shape.py:384]: If you're concerned about that, then you can use `aet.lvector.filter_variable`; that will normalize the type and emit an informative error.

Comment (michaelosthege) [reply aesara/tensor/shape.py:384]: `filter_variable` doesn't convert the integer types either: `TypeError: Cannot convert Type TensorType(int32, (True,)) (of Variable TensorConstant{(1,) of 2}) into Type TensorType(int64, vector).`

**Comment by michaelosthege** [comment]
Okay, so I think that's as concise as it gets. Delegating type casting and raising errors on another functino didn't work for me, because int32 inputs also blew up. The type check was just nicer to work with after all.

I tried to align the C and Python exception messages, but burnt 2 hours trying to string-format the shape array for the exception message. (Now I get why that wasn't done before...)

**Review by brandonwillard — APPROVED** [review]
This reminded me that `aet.as_tensor_variable` doesn't have a `dtype` option.  We should add that.

**Pull request merged**